src/bluesky/tests/test_cross_section_image.py

```diff
@@ -0,0 +1,248 @@
+    with patch("matplotlib.pyplot.figure") as mock_fig:
+        mock_figure_instance = MagicMock(spec=plt.Figure)
+        mock_subplot = MagicMock(spec=plt.Axes)
+        mock_figure_instance.add_subplot.return_value = mock_subplot
+        mock_subplot.get_figure.return_value = mock_figure_instance
+        mock_fig.return_value = mock_figure_instance
+        yield mock_fig
+
+
+@pytest.fixture
+def mock_cross_section():
+    return CrossSection(figure=MagicMock())
+
+
+@pytest.fixture
+def live_image(mock_cross_section):
+    return LiveImage(
+        "test_field",
+        cmap=Colormap(colormaps["magma"]),
+        norm=Normalize(),
+        interpolation=InterpolationEnum.NONE,
+        cross_section=mock_cross_section,
+    )
+
+
+class TestAutoRedrawDecorator(TestCase):
+    def setUp(self):
+        class TestClass:
+            def __init__(self):
+                self._fig = MagicMock()
+                self._auto_redraw = True
+
+            @auto_redraw
+            def some_function(self, x, y):
+                return x + y
+
+            def _update_artists(self):
+                pass
+
+            def _draw(self):
+                pass
+
+        self.obj = TestClass()
+
+    def test_function_call_and_return_value(self):
+        result = self.obj.some_function(2, 3)
+        self.assertEqual(result, 5, "The decorated function should return the sum of inputs")
+
+    def test_redraw_called_when_auto_redraw_true(self, mock_cross_section: CrossSection):
+        mock_cross_section.draw_idle()
+        self.obj.some_function(2, 3)
+        self.obj._fig.canvas.assert_not_called()
+        self.obj._update_artists.assert_called_once()
+        self.obj._draw.assert_called_once()
+
+    def test_redraw_not_called_when_auto_redraw_false(self):
+        self.obj._auto_redraw = False
+        self.obj.some_function(2, 3, force_redraw=False)
+        self.obj._update_artists.assert_not_called()
+        self.obj._draw.assert_not_called()
+
+    def test_redraw_called_with_force_redraw_true(self):
+        self.obj._auto_redraw = False  # Make sure _auto_redraw is false
+        self.obj.some_function(2, 3, force_redraw=True)
+        self.obj._update_artists.assert_called_once()
+        self.obj._draw.assert_called_once()
+
+    def test_no_canvas_no_redraw(self):
+        # Simulate the absence of a canvas
+        self.obj._fig.canvas = None
+        self.obj.some_function(2, 3)
+        self.obj._update_artists.assert_not_called()
+        self.obj._draw.assert_not_called()
+
+
+class TestCrossSection(TestCase):
+    def setUp(self):
+        self.figure = MagicMock(spec=Figure)
+        self.cross_section = CrossSection(figure=self.figure, colormap="gray", auto_redraw=True)
+        self.cross_section._imdata = MagicMock(shape=(100, 100))  # Dummy image data
+        self.cross_section._cursor_position_cbs.append(MagicMock())  # Add mock callback
+
+    @patch("mpl_toolkits.axes_grid1.make_axes_locatable")
```

Comment:
You're right that you need to patch this function but:
* When you're patching you actually need to patch where the function has been imported to, not where it was imported from
* You need to patch this before `CrossSection` is instantiated in `setUp`
* Annoyingly this is a bit fiddly https://stackoverflow.com/questions/15821465/how-to-properly-use-mock-in-python-with-unittest-setup

The following gives an instantiatable `CrossSection` and passes this test:
```python
class TestCrossSection(TestCase):
    def setUp(self):
        patch_divider = MagicMock()
        patch_divider.append_axes.return_value.plot.return_value = (0, ) # Patch for _ln_v and _ln_h
        patch("bluesky.callbacks._mpl_image_cross_section.make_axes_locatable", return_value=patch_divider).start()
        self.figure = MagicMock()
        self.cross_section = CrossSection(figure=self.figure, colormap="gray", auto_redraw=True)
        self.cross_section._imdata = MagicMock(shape=(100, 100))  # Dummy image data
        self.cross_section._cursor_position_cbs.append(MagicMock())  # Add mock callback

    def test_move_cb_outside_axes(self):
        event = MagicMock()
        event.inaxes = None
        self.cross_section._move_cb(event)
        self.cross_section._cursor_position_cbs[0].assert_not_called()
```

Reply:
ah I forgot about this